More vehicle part descriptions

mlangsdorf opened the pull request:
Expand some vehicle part descriptions based on feedback from reddit and discourse.

Refactor the code that prints the vehicle part description so it can be used it more places than just the install menu.

Print the descriptions for all vparts on the current tile in the general vehicle interaction display, using the window panel that was otherwise blank

ZhilkinSerg commented:
# Minor issues (can be addressed in later PRs):

- pressing `>`, `<`, `TAB` or `ENTER` clears window with parts descriptions;
- there are no keybindings to scroll window with part descriptions.

mlangsdorf commented:
I can add a commit to fix the minor issues.

I though fold_and_print would handle the scrolling automatically, but if it doesn't, I can copy the logic from print_parts_list.

ZhilkinSerg commented:
> I though fold_and_print would handle the scrolling automatically, but if it doesn't, I can copy the logic from print_parts_list.

Maybe it does, but there is no keybinding for it and description for more than 3 parts are garbled.

![image](https://user-images.githubusercontent.com/16213433/43156712-90bb01a6-8f83-11e8-8fa8-1e7b994cb21d.png)


![image](https://user-images.githubusercontent.com/16213433/43156700-8c3bb4d6-8f83-11e8-9756-54b34b632349.png)

mlangsdorf commented:
Right, I agree there is a problem. I was just explaining why I didn't code it better - because I thought I had. Ah, well.
Working on a fix, and poking at it more already let me fix the x for y swap that was making the word wrapping so cramped.

mlangsdorf commented:
That was fun.

The descriptions list should be truncated to the window size now, with a message indicating the scroll keys. Hitting TAB, BACKTAB, <, and >, should all do reasonably sensible things and the vehicle
parts description list won't disappear when you're scrolling through the fuel list.

mlangsdorf commented:
Okay, this commit has grown to ~950 lines. I can split off the data/json/vehicle_parts changes if people would like, but the src changes are fairly tightly coupled.

AMurkin commented on data/json/vehicleparts/engineering.json:
```diff
@@ -28,7 +28,7 @@
     "difficulty": 2,
     "durability": 100,
-    "description": "An extendable cantilever crane.  Can be used to jack up another vehicle to change the other vehicle's wheels, or lift a heavy component like an engine from another vehicle.",
+    "description": "An extendable cantilever crane.    If it is in your line of sight and within two tiles of another vehicle, you will automatically use if when you want jack up the other vehicle to change the its wheels, or lift a heavy component like an engine that you are installing in or removing from the other vehicle.",
```
Too many spaces here and in two next changes.

ZhilkinSerg commented:
> The descriptions list should be truncated to the window size now, with a message indicating the scroll keys. Hitting TAB, BACKTAB, <, and >, should all do reasonably sensible things and the vehicle
parts description list won't disappear when you're scrolling through the fuel list.

Scrolling vehicle part description window with `>` does not have upper limit, so you can scroll through empty pages indefinitely - only `< More parts here..` header will be shown:

![image](https://user-images.githubusercontent.com/16213433/43186205-af0589a6-8ff6-11e8-8396-b824bce88691.png)

Also `ENTER` (`Confirm Choice`) still clears vehicle part description window.

> Okay, this commit has grown to ~950 lines. I can split off the data/json/vehicle_parts changes if people would like, but the src changes are fairly tightly coupled.

No need to split yet.

mlangsdorf commented:
Fixed the scroll-off issue. It's approximate, but seemed adequate in my testing.
Fixed the no-display on RETURN issue - now vehicle part descriptions pop up whenever the w_msg isn't being used for something more important.

Fixed the typos in the JSON - thanks for the review.

ZhilkinSerg commented:
> Fixed the scroll-off issue. It's approximate, but seemed adequate in my testing.

The bug is still on, but only for tiles with little vehicle parts (when all descriptions go to first page).

Here is first page - it does not contain `More parts here... >` label:

![image](https://user-images.githubusercontent.com/16213433/43264479-feaab38e-90ee-11e8-9015-11fa9a396865.png)

But I can still press `>` and proceed to empty pages:

![image](https://user-images.githubusercontent.com/16213433/43264520-13fbfd9c-90ef-11e8-9218-752ed09b5485.png)

And if I go to tile with multiple parts (which descriptions do not fit on one page):

![image](https://user-images.githubusercontent.com/16213433/43264605-45356812-90ef-11e8-87f6-924f794aec6d.png)

I can only proceed to some ultimate page with description of last part:

![image](https://user-images.githubusercontent.com/16213433/43264608-48fe7b28-90ef-11e8-98ad-d5ae42c4a630.png)

mlangsdorf commented:
More fun!
Okay, you can scroll down until the remaining descriptions fit in the display and no further. You can scroll up to the first part. Moving resets the scroll limits.
I think this works correctly.

ZhilkinSerg commented:
Looks good now!
